File: java/src/main/java/com/github/mostroverkhov/r2/java/R2Server.java
package com.github.mostroverkhov.r2.java;

import com.github.mostroverkhov.r2.core.internal.MimeType;
import com.github.mostroverkhov.r2.core.internal.acceptor.ServerAcceptor;
import com.github.mostroverkhov.r2.core.R2ServerFluentBuilder;
import io.rsocket.Closeable;
import io.rsocket.ConnectionSetupPayload;
import io.rsocket.RSocket;
import io.rsocket.RSocketFactory.ServerRSocketFactory;
import io.rsocket.RSocketFactory.Start;
import io.rsocket.transport.ServerTransport;
import org.jetbrains.annotations.NotNull;
import reactor.core.publisher.Mono;

public class R2Server<T extends Closeable> extends R2ServerFluentBuilder<
        ServerRSocketFactory,
    JavaServerAcceptorBuilder,
        ServerTransport<T>,
        Start<T>> {

    @Override
    public Start<T> transport(ServerTransport<T> transport) {
        assertState();
        JavaServerAcceptorBuilder acceptorBuilder = new JavaServerAcceptorBuilder();
        ServerAcceptor<ConnectionSetupPayload, RSocket,Mono<RSocket>> acceptor =
                getAcceptorConfigurer()
                        .invoke(acceptorBuilder)
                        .build();

        return getServerRSocketFactory()
                .addConnectionPlugin(setupInterceptor())
                .acceptor(() -> acceptor::accept)
                .transport(transport);
    }

    @NotNull
    private SetupInterceptor setupInterceptor() {
        return new SetupInterceptor(MimeType.INSTANCE);
    }
}

File: android/src/main/java/com/github/mostroverkhov/r2/android/R2Server.kt
package com.github.mostroverkhov.r2.android

import com.github.mostroverkhov.r2.core.internal.MimeType
import com.github.mostroverkhov.r2.core.internal.responder.ServerFluentBuilder
import io.rsocket.android.Closeable
import io.rsocket.android.ConnectionSetupPayload
import io.rsocket.android.RSocket
import io.rsocket.android.RSocketFactory.ServerRSocketFactory
import io.rsocket.android.RSocketFactory.Start
import io.rsocket.android.SocketAcceptor
import io.rsocket.android.transport.ServerTransport

class R2Server<T : Closeable> : ServerFluentBuilder<
        ServerRSocketFactory,
        AndroidAcceptorBuilder,
        ServerTransport<T>,
        Start<T>>() {
    override fun transport(transport: ServerTransport<T>): Start<T> {
        assertState()
        val configurer = acceptorConfigurer!!
        val rsocketFactory = serverRSocketFactory!!

        val acceptorBuilder = AndroidAcceptorBuilder()
        val acceptor = configurer(acceptorBuilder).build()
        return rsocketFactory
                .addConnectionPlugin(setupInterceptor())
                .acceptor { adapt(acceptor) }
                .transport(transport)
    }

    private fun setupInterceptor() = SetupInterceptor(
            MimeType.dataType,
            MimeType.metadataType)

    private fun adapt(acceptor: Acceptor): SocketAcceptor = object : SocketAcceptor {
        override fun accept(setup: ConnectionSetupPayload,
                            sendingSocket: RSocket) = acceptor.accept(setup)
    }
}
